### PR TITLE
Migrate from CPPUnit to gtest

### LINKED ISSUE #3439: Switch Testing Framework from CppUnit
**Is your feature request related to a problem? Please describe.**

This is not strictly a problem, but given the fact that we currently don't have a lot of tests, it could be a good opportunity to switch to a more mature and feature complete test framework.

**Describe the solution you'd like**

I'd like to have a cpp testframework with:
 - good documentation
 - wide adoption
 - ease of use (tests should be easy to write)
 - being compatible with mocking frameworks

**Describe alternatives you've considered**

Possible Solutions are:
 1. [gtest](http://google.github.io/googletest/) with [gmock](http://google.github.io/googletest/gmock_for_dummies.html)
 2. [catch2](https://github.com/catchorg/Catch2) with gmock (example on how this might look like https://github.com/ecokeley/catch_gmock_integration)

For 1. one could also allow for BDD-Style testing by including [cppbdd](https://github.com/michaelvlach/cppbdd).

### PR DESCRIPTION
**Pull request by idotobi**
- include gtest to cmake files and remove cppunit
 - include gtest in azure-pipelines and remove cppunit
 - move tests from test/ to gtest/unit_tests/ and adapt to gtest
 - add two testcases from XojPreviewExtractorTest that were previously
   skipped and fix them

Resolves #3439

### CODE REVIEW

**Comment by idotobi** [comment]
>     * It really doesn't matter, but I find `test` more convenient.

Makes sense. Initially I thought there would be a period where both testing systems would be available in parallel, but now that all CPPUnit stuff is already migrated, I think `test` is the more canonical choice.

**Comment by idotobi** [comment]
@LittleHuba can I simply call `/azp run create-installers` to check if the installer pipeline works as well or does this have any side effects I should keep in mind?

**Comment by LittleHuba** [comment]
That should work without any side effects. You'll just have to make sure that your branch originates from a commit that is later than the one where I added this feature. Or just try it and if it fails rebase.

**Comment by idotobi** [comment]
The debian build seems to fail because the image [xournalpp/debian-latest-sudo](https://hub.docker.com/r/xournalpp/debian-latest-sudo) is still on buster ([cmake 3.13.4-1](https://packages.debian.org/buster/cmake)) and not on bullseye ([cmake 3.18.4-2](https://packages.debian.org/bullseye/cmake)). 

Judging by the [Dockerfile](https://github.com/xournalpp/debian-latest-sudo/blob/master/Dockerfile) simply rebuilding the image and pushing to DockerHub should be enough.

@LittleHuba: I don't have the credentials to upload a new docker image. Could you update the image?

**Comment by LittleHuba** [comment]
I updated the image. We should now be on buster.

**Comment by LittleHuba** [comment]
You'll have to rebase once #3547 is merged. I'll merge as soon as it builds successfully.

**Comment by LittleHuba** [comment]
The requested installers were generated successfully. The installers are available at:

- [Windows](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY0MzMvYXJ0aWZhY3ROYW1lL3dpbmRvd3M1/content?format=zip)
- [macOS](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY0MzMvYXJ0aWZhY3ROYW1lL21hY09T0/content?format=zip)
- [Debian](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY0MzMvYXJ0aWZhY3ROYW1lL2xpbnV4LWRlYmlhbi1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 18.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY0MzMvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0xOC4wNC1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 20.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY0MzMvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0yMC4wNC1wYWNrYWdlcw2/content?format=zip)

Please be aware that these links are only valid for a limited time (~30 days).

**Comment by idotobi** [comment]
@LittleHuba , @Febbe : If there are no objections I would *squash an merge* this tomorrow.

**Comment by Febbe** [comment]
I think you can prepare the squash and merge it. But honestly, I don't have time to take a last look

**Comment by LittleHuba** [comment]
Same from my side. I have no free time to check this prior to the 15th. I guess we'll have to trust in your work. But I can do that with a good conscience. You may go ahead and merge this if you think it is good to go.

File: CMakeLists.txt [hunk 1]
```diff
@@ -24,6 +24,12 @@ set_directory_properties(PROPERTIES EP_BASE "${PROJECT_BINARY_DIR}/external")
 
 add_definitions(-D_USE_MATH_DEFINES)
 
+# Check for OS
+if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+    set(MACOSX TRUE)
```
Comment (personalizedrefrigerator) [inline CMakeLists.txt:29]: A quick note for the future (if we decide to add FreeBSD support):

Table from gitlab.kitware.com: CMAKE_SYSTEM_NAME is set to Darwin on (seemingly) only MacOS/iOS

https://gitlab.kitware.com/cmake/community/-/wikis/doc/tutorials/How-To-Write-Platform-Checks#cmake-system

It looks like this won't be an issue if we decide to add FreeBSD support &mdash; `${CMAKE_SYSTEM_NAME}` is `FreeBSD` on FreeBSD.

File: CMakeLists.txt [hunk 2]
```diff
@@ -140,13 +146,21 @@ pkg_check_modules(SNDFILE REQUIRED "sndfile >= 1.0.25")
...
+        URL https://github.com/google/googletest/archive/refs/tags/release-1.11.0.zip
+    )
+	# Prevent reloading if already downloaed
+	set(FETCHCONTENT_UPDATES_DISCONNECTED ON)
+    # For Windows: Prevent overriding the parent project's compiler/linker settings
+    set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
+    FetchContent_MakeAvailable(googletest)
 	enable_testing()
-endif (ENABLE_CPPUNIT)
+endif (ENABLE_GTEST)
```
Comment (personalizedrefrigerator) [inline CMakeLists.txt:163]: **Potential follow-up PR:** Doing the above only if [`find_package(GTest CONFIG)`](https://github.com/google/googletest/blob/master/googletest/README.md#incorporating-into-an-existing-cmake-project) fails might be nice. (I can try adding this!)

File: test/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+* replace `CPPUNIT_ASSERT_EQUAL`  
+  with `EXPECT_EQ`
+
+## How to add a new test program
+
+First why: This would make sense in case you have a very different testing usecase than with unit-tests and you might want to collect code coverage separately. Otherwise, adding to unit-tests is way simpler and preferred as  `gtest_discover_tests` does all the book keeping for you then.
+
+* create a new folder for your test program
+* add the definition of your test program to `gtest/CMakeLists.txt`
+* add the test program to
```
Comment (personalizedrefrigerator) [inline test/README.md:42]: Please finish this line!

File: test/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+TEST(HelloTest, BasicAssertions) {
+  // Expect two strings not to be equal.
+  EXPECT_STRNE("hello", "world");
+  // Expect equality.
+  EXPECT_EQ(7 * 6, 42);
+}
+```
+as taken from the official [docs](http://google.github.io/googletest/quickstart-cmake.html).
+
+As all `gtest/unit_tests` are build with a dependency on `xournalpp-core` you can include any file from `src` as you would in the main code.
```
Comment (personalizedrefrigerator) [inline test/README.md:24]:
```suggestion
As all `gtest/unit_tests` are built with a dependency on `xournalpp-core` you can include any file from `src` as you would in the main code.
```

**Review by personalizedrefrigerator — APPROVED** [review]
This looks good to me! I have a few (very minor) comments, but I would be fine with merging this as-is!

Note that this review is just based on a quick look through the diff. I haven't tried building this.

Thank you for your work on this!

File: test/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  Reason: image not found
+```
+
+then you have some fun `@rpath` problem.
+The `gtest/CMakeLists.txt` already contains a solution for this, so that is a good start for debugging if something related to MacOSX build failures would come up.
+
+Apart from that resources I found useful:
+ * [Understanding RPATH (with CMake)](https://dev.my-gate.net/2021/08/04/understanding-rpath-with-cmake/)
+ * [RPATH handling](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling)
+
```
Comment (personalizedrefrigerator) [inline test/README.md:67]:
```suggestion

## Problems running `make test`

If CMake is generating UNIX Makefiles and `make test` fails with  the error, `Unable to find executable: test-units_NOT_BUILT`, make sure,
 1. `make test-units` has been run (to compile the test executable)
 2. The project was configured with `-DENABLE_GTEST=ON` (e.g. by configuring with `cmake .. -DENABLE_GTEST=ON`)

```

I was having trouble running the test suite on Manjaro. Running `make test-units` before running `make test` successfully ran the tests. Some users may find the above addition helpful.

File: test/README.md (hunk 4 above)
Comment (idotobi) [reply test/README.md:24]: :sweat_smile:  good catch.

File: test/README.md (hunk 5 above)
Comment (idotobi) [reply test/README.md:67]: Good point I'll add that. I always just `build & run` the test-units executable. Hence, I didn't notice.
Seems this is a known problem of cmake https://gitlab.kitware.com/cmake/cmake/-/issues/8774 ([stackoverflow](https://stackoverflow.com/questions/733475/cmake-ctest-make-test-doesnt-build-tests)).

**Comment by idotobi** [comment]
@personalizedrefrigerator : Thanks for giving it a spin, and double checking on the README :pray:  I incorporated all your comments.

**Comment by bhennion** [comment]
@idotobi Whenever you have time, could you update the build README's to use `gtest` instead of `cppunit`? Those pages still refer to `cppunit` as a build dependency:
[Linux build](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md)
[Windows build](https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md)

The [MacOs build](https://github.com/xournalpp/xournalpp/blob/master/readme/MacBuild.md) page does not mention tests at all for the moment.

**Comment by idotobi** [comment]
@bhennion : Thank you for noting I created  #3560 for this. I think I'll have time tomorrow.

**Pull request merged**